### PR TITLE
GLT-3131: added caching options

### PR DESCRIPTION
**Pull request by djcooke**

### CODE REVIEW

File: pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    Date afterDate = after == null ? null : Date.from(after.toInstant());
+
+    return samples.stream()
+        .filter(
+            sample -> {
+              if (archived != null && !archived.equals(sample.getArchived())) {
+                return false;
+              }
+              if (normalizedProjects != null
+                  && normalizedProjects.stream()
```
Comment (apmasell) [inline pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java]: I think these can be reduced to a `!normalizedX.contains(sample.getX())`

File: pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+  private static final Logger log = LoggerFactory.getLogger(Cache.class);
+
+  private static final Histogram cacheUpdateTime =
+      Histogram.build()
+          .name("pinery_cache_update_time")
+          .help("Time to update the cache (in seconds)")
+          .buckets(60, 180, 300, 600, 900, 1200, 1800, 3600)
+          .register();
+  private static final Gauge cacheUpdateFailures =
```
Comment (apmasell) [inline pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java]: Can you also include a gauge with the timestamp of when the cache was updated? There's a `setToCurrentTime()` method.

File: pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java (hunk 1 above)
Comment (djcooke) [reply pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java]: changed

File: pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java (hunk 2 above)
Comment (djcooke) [reply pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java]: added

**Comment by djcooke** [comment]
Updates:

* scheduled updates can be disabled by setting an interval of zero or less. When scheduled updates are disabled, there's no automatic update on startup either
* if cache is already updating when update is called, the second update now blocks until the first completes. Previously, the second would return before the update completed in the first call, which meant that there was no guarantee the cache was done refreshing when update returned
* properties are now required; no defaults
* an error status is now returned from `/updatecache` when caching is disabled

**Pull request merged**